### ISSUE TITLE
watch Option for SAM Build command

### ISSUE THREAD

**Issue by jfuss** [awslabs/aws-sam-cli#921]
### Describe your idea/feature/enhancement

When using `sam build` to build functions that will later be invoked through `sam local [invoke|start-lambda|start-api]`, you need to run `sam build` before each invoke. The reason for this requirement is because the build command places artifacts (code + dependencies + updated template) into a build folder. The `sam local [invoke|start-lambda|start-api]` use these built templates (this has updated locations to the built function code) by default but there is no way to update the built code without running `sam build` again.

### Proposal

For the initial implementation of `sam build`, the `--watch` option [was out of scope](https://github.com/awslabs/aws-sam-cli/blob/develop/designs/sam_build_cmd.md#out-of-scope). The proposal is to add support for this option to all customers to enable building when source code changes.

An alternative could be to make the `sam local` suite of commands build the function during invoke. The main concern with this is the speed at which you can now invoke (other build enhancements might be a requirement for this, speed, incremental builds, etc).

### Workaround

All of the `sam local` command allow 'reloading' of the function code. This means for a given template, if you update the contents of a functions Code/CodeUri, SAM CLI will mount this updated content. So the current option for making interactions easier is to have two terminal windows open, one that builds and one that you local invoke with.

**Comment by alxx** [comment]
To me this only happens using Ruby once I've added a gem into the equation, with my functions having `Gemfile`s in their respective directories. Then I'm forced to build using `--use-container` parameter, which takes a lot of time; simply running `sam build` doesn't work because the app doesn't find the gems.

If I don't use gems at all, then I have live reloading; but once I need gems, everything only works by rebuilding inside a container. And since that takes about a minute, it makes SAM pretty much unusable for me. I can't build every time I make a small change, it just takes too long.

**Comment by AmillerSC** [comment]
Is there any progress on this? Having to build every time for the most minor of code changes is not ideal.

**Comment by sliedig** [comment]
I'd like to see a slightly different take on this proposal. Rather than having a continual build function through `--watch` option, only build functions that have changed since the last build. I often find myself having to rebuild functions that haven't been modified, or just making changes to the template file, making the feedback loop unnecessarily long.

**Comment by tamvm** [comment]
> To me this only happens using Ruby once I've added a gem into the equation, with my functions having `Gemfile`s in their respective directories. Then I'm forced to build using `--use-container` parameter, which takes a lot of time; simply running `sam build` doesn't work because the app doesn't find the gems.
> 
> If I don't use gems at all, then I have live reloading; but once I need gems, everything only works by rebuilding inside a container. And since that takes about a minute, it makes SAM pretty much unusable for me. I can't build every time I make a small change, it just takes too long.

Hi @alxx, 

So what is your workaround for that? Or can you share your experience on that? 

Thank you,

**Comment by mims92** [comment]
I don't like to compare but Chalice as this feature if I'm not mistaken ;)

**Comment by justengland** [comment]
Assuming this is hard to build the watch option since this has been open for the better part of a year. It seems an easier option would be to use a docker volume to share your code rather than requiring a build, this way you do not have to have an actual watch.

**Comment by dgomesbr** [comment]
For a workaround, you're able to simply delete the ``.aws-sam`` folder at the root of your project and get back invoking your function.

ref: ``rm -rf .aws-sam/*``at root.

**Comment by dz902** [comment]
Lacking of a `watch` mechanism is eating much of my time (and laptop battery). Would love to have this feature. Especially we are so full-hearted dedicate ourselves to serverless thus creating lots of APIs.

**Comment by serverhorror** [comment]
> For a workaround, you're able to simply delete the `.aws-sam` folder at the root of your project and get back invoking your function.
> 
> ref: `rm -rf .aws-sam/*`at root.

That doesn't work for me. Running in python3.8 now and it doesn't pick up the packages in `requirements.txt`.

**Comment by codecounselor** [comment]
If you're using a **Node runtime** you may want to consider using Webpack to build bundles for each function instead of `sam build`.  Then you can use `webpack -w` to have it watch for changes which `sam local` picks up.

I followed this very excellent write-up: https://dev.to/elthrasher/managing-multiple-functions-with-aws-sam-and-webpack-1581
I found https://www.npmjs.com/package/aws-sam-webpack-plugin also but preferred the simplicity of the former setup.

The relevant config:

### webpack.config.js
```javascript
module.exports = {
	// Specify the entry point for our app
	entry: {
		functionA: path.join(__dirname, 'src/functionA.js'),
		functionB: path.join(__dirname, 'src/functionB.js'),
	},
	// Specify the output file containing our bundled code
	output: {
		libraryTarget: 'commonjs2',
		path: `${__dirname}/build`,
		filename: '[name].js',
	},
	mode: process.env.NODE_ENV === 'dev' ? 'development' : 'production',
	target: "node", // Let webpack know to generate a Node.js bundle
}
```

### template.yaml
```yaml
  PingFunction:
    Type: AWS::Serverless::Function
    Properties:
      CodeUri: build/
      Handler: functionA.lambdaHandler
      Runtime: nodejs12.x
```

1. `npx webpack -w`
1. `sam local start-api`

**Comment by fakekamrankhan** [comment]
I just used nodemon for this. Going to put this here in case it would make anyone else's life easier:

```
npm i -g nodemon
nodemon --exec sam build
```

Run this in the root folder in one terminal and run `sam local start-api` in another. nodemon should build the application automatically on any code changes.

**Comment by eamarce** [comment]
Hey guys,

So, we created a small npm package based in nodemon (samwatch):

https://www.npmjs.com/package/samwatch

https://github.com/mxitgo/samwatch

The way it works is, it copies your js/json files as you save them from the source folder to the corresponding .aws-sam/build folder (as long as the names are the same for your lambda and the code uri in template.yaml).

That is useful to see your changes reflected on the fly if you are running sam local start-api, so you don't have to run sam build after every change.

On the other hand, if the corresponding file copy is not within the .aws-sam/build folder already, the package will trigger a sam build (unless you use n parameter).

The package has helped us get a hot reload feeling for our local lambda function development with sam.

I hope anyone can find it useful.

Thanks!

**Comment by butsjoh** [comment]
My usage is also with ruby code so i am using the live reloading as follows currently without actually using sam build. Not sure though if this is recommend or bad practice so would like to get feedback on it. 

Let say i have the following folder structure with 2 functions:

```
lambda_example/
|--hello
||--app.rb
||--Gemfile
|--hello_other
||--app.rb
||--Gemfile
|--template.yaml
```

instead of using sam build i just do: (i do this from the root directory lambda_example)

```
# To install new gems for hello (only need to run this when you update the Gemfile or when you don't have a Gemfile.lock yet)
bundle install --gemfile=hello_world/Gemfile

# Now the following installs the gems into the vendor bundle inside the function directory
# similar what sam build does actually if you would look in the .aws-sam folder
bundle install --gemfile=hello_world/Gemfile --deployment --path vendor/bundle
```

Same thing then for the other function

```
bundle install --gemfile=hello_other/Gemfile
bundle install --gemfile=hello_other/Gemfile --deployment --path vendor/bundle
```

Now if you then issue sam local start-api and you make changes to the code then the live reloading stuff should be working without actually needing to build each time. Ofcourse this only probably works if you don't need to build native extensions for gems but esentially you only need to call the bundle install commands above if you change gem dependencies which should not happen that often.

The samwatch thing is cool but does not work for ruby cause it only monitors js or json files it seems. So maybe add an extra option in your cli @eamarce to specify which files to watch?

**Comment by paulmowat** [comment]
Now that warm containers has been delivered under #2383 . Is there anyway this can be prioritized? It would help make that dev experience so much better.

**Comment by anubhavmalik** [comment]
Its such a painful process running sam build even for a small change such as a console.log(). Any good workaround? The solutions above (webpack and samwatch) force a folder structure which I am not willing to do.

**Comment by mims92** [comment]
@anubhavmalik only thing is to modify the code in the build folder.

**Comment by dror-g** [comment]
**such a pain**... maybe instead of a "watch" option, which will still need to build on every change and slow down development we could mount the source folder in the container?

**Comment by ppejovic** [comment]
In the case of lambda [container images](https://docs.aws.amazon.com/lambda/latest/dg/lambda-images.html) it looks like a `sam build` is always required as the build output is an image rather than code being pushed to the `.aws-sam` directory. I have a working docker compose setup that spins up two services; one serving the API with `sam local start-lambda` and the other that rebuilds the images with `sam build` on code changes using `nodemon`. This works, and will likely still work with lambdas being packaged as `.zip`s, but it still feels like a heavyweight solution:

https://github.com/ppejovic/sam-docker-template

**Comment by 9oelM** [comment]
> In the case of lambda [container images](https://docs.aws.amazon.com/lambda/latest/dg/lambda-images.html) it looks like a `sam build` is always required as the build output is an image rather than code being pushed to the `.aws-sam` directory. I have a working docker compose setup that spins up two services; one serving the API with `sam local start-lambda` and the other that rebuilds the images with `sam build` on code changes using `nodemon`. The feedback loop could be quicker but it works fairly well.

I've been actually doing this and I am curious if this will produce many local images as it builds again. Is it going to do that? Seems like that's what's happening to mine.

**Comment by jhechtf** [comment]
I remember using a previous version of the SAM cli and being able to just run the `sam local start-api` and so long as I didn't need to add a new endpoint to the `template.yml` file the code was reflected automatically, no need for `sam build` every single time. I even have a repository setup with a nodemon script set explicitly to watch only for changes in the `template.yml` file so that I wouldn't forget to restart it if I modified the template file. 

I'm wondering why this behavior was changed? Now I'm honestly considering not using the SAM cli because this workflow is rather annoying.

**Comment by mims92** [comment]
Maybe a simple solution to that: make a symbolic-link (ln -s) from the source code to the build folder?

Something like:

```
sam build ...
ln -s src/* .aws-sam/build/FunctionName
```

**Comment by jfuss** [comment]
@jhechtf That behavior still holds true the difference is when you use build, SAM CLI places the artifacts into another directory with an updated template. When running `sam local start-api`, it will read the "built" template. So you can get the same behavior through `sam build` in a different terminal or not using `sam build` at all and therefore the "source" template is always read. 

As a note, we do prioritize off of reactions. So more reactions do help us in deciding priority for outstanding feature requests from the community.

**Comment by jhechtf** [comment]
@jfuss I've seen that mentioned before, about simply not using "sam build", but when I remove the `.aws-sam` directory and attempt to run `sam local start-api` I receive errors stating that it can't find any node modules, e.g. I use the `@aws-sdk/client-dynamodb` and without running `sam build` I get the error of `Error: Cannot find module '@aws-sdk/client-dynamodb'`

**Comment by jfuss** [comment]
@jhechtf Is the node modules within the CodeUri of the Function? If not, you need to install your dependencies into the CodeUri location. SAM CLI mounts the CodeUri into the container, which means all the code needs to be under that directory. This is what `sam build` does for you, it just does not mutate the source and instead places everything under `.aws-sam/`

**Comment by jhechtf** [comment]
@jfuss the template that I am using is the nodejs14.x generated by the sam cli, which does not include any CodeUri properties. I'll try to add that in, but considering the code gets generated in the `<project root>/src/*`  and the node_modules are at `<project root>` I'm not entirely sure that's going to work out for me.

**Comment by jfuss** [comment]
@jhechtf Happy to help but let's spin this off into a different issue. This will distract from the original intent of this issue/request.

**Comment by memeyeop** [comment]
Check it.

- template.yaml 예시
    - 사용하는 라이브러리를 로컬경로의 Layer로 설정

```yaml
CommonLayer:
    Type: AWS::Serverless::LayerVersion
    Properties:
        LayerName: common-dependencies
        Description: Common dependencies
        ContentUri: layers/common/
        CompatibleRuntimes:
            - python3.8
        RetentionPolicy: Retain

HelloWorldFunction:
    Type: AWS::Serverless::Function # More info about Function Resource: https://github.com/awslabs/serverless-application-model/blob/master/versions/2016-10-31.md#awsserverlessfunction
    Properties:
        CodeUri: functions/consult/
        Handler: app.lambda_handler
        Runtime: python3.8
        Layers:
          - !Ref CommonLayer
```

- 로컬경로에 requirements.txt에 명시된 라이브러리 설치
    - 경로 예시: layers/common/python 아래에 설치 (최상위 폴더이름은 "python" 필수)

```bash
$ cd layers/commons/
$ pip install -r ../상대경로/requirements.txt -t ./python
```

- 로컬 실행 시 템플릿 파일 직접 명시 (프로젝트 내 build 폴더는 없어야 함)
```bash
$ sam local start-api -t template.yaml --skip-pull-image
```

**Comment by metaskills** [comment]
## Summary

I am 100% NOT a fan of this watch idea. Editing code and hitting refresh would still take a long time. Perhaps the solution to the problem is rethinking what is needed. Perhaps... can we have a `start-api` option that does a simple API Gateway => Lambda simulation similar to Zip which would do the following so we can have code reloading for containers?

- Does not need `sam build` to run.
- Assumes the host has everything needed like platform, packages, system deps, etc.
- Avoids using a "runtime" container with docker mounts. Because above.

I do not know if this is a fair ask of the SAM team. Maybe I need the work y'all have done inside of SAM to be a smaller components to assemble. Maybe we have this already and I have not found a good way to do just that... for my needs?

## Context & Hack Attempts

For both Ruby, Node, etc, we use the SAM build image as a Docker in Docker development image. A good example of this in practice for Ruby/Rails is here. https://lamby.custominktech.com/docs/quick_start. When doing this, we can assume the local "host" (docker dev image) has everything we need, system dependencies, dev code which can be edited, etc. All we need is a simple API Gateway => Lambda emulator.

For Rails, we completely side step this issue by using the built-in `./bin/rails server` and the need for API Gateway => Lambda handler integration because that is tested in abstract at another layer. See here: https://github.com/customink/lamby

Today I was coming back to an old Node zip function I wrote a few years ago and wanted to move it to our latest patterns which includes SAM dev containers, production package type image, etc and was super frustrated that SAM does not have a  way to support start-api with container images. I tried a few hacks. First, I tried a little bit of @mims92 recommendation by leaning into the need to do `sam build` (which technically I do not need because I an assume the host has everything) to trick `start-api` into thinking it had what it needed. Along with forcing the host because Docker in Docker with compose, etc. Again, see Lamby Rails cookiecutter. Still, I could not edit my code and hit refresh in the browser.

```shell
if [ ! -d "./.aws-sam" ]; then
  sam build \
    --parameter-overrides "StageEnv=development"
  mkdir  ./.aws-sam/build/Lambda
  pushd ./.aws-sam/build/Lambda
  ln -s ../../../src src
  popd
fi

sam local start-api \
  --host '0.0.0.0' \
  --port 3031 \
  --container-host 'host.docker.internal'
```

I then had the crazy idea maybe I could use the RIE (https://docs.aws.amazon.com/lambda/latest/dg/images-test.html) built into my image. So I tried adding this to my `Dockerfile-build` (dev & build image) along with a change to my local `./bin/server` script. But I quickly learned this is too LOW LEVEL and not going to work for me. I need the API Gateway integration.

```dockerfile
RUN curl -Lo /usr/local/bin/aws-lambda-rie https://github.com/aws/aws-lambda-runtime-interface-emulator/releases/latest/download/aws-lambda-rie \
    && chmod +x /usr/local/bin/aws-lambda-rie
```

```shell
docker-compose run \
  --entrypoint "/usr/local/bin/aws-lambda-rie --log-level debug" \
  -p 9001:3031 \
  node-container-test
```

For this hack I tried to see if I could have a development only version of my (`template-dev.yaml`) to use just with `start-api` to see if I could get some simple Zip behavior in my dev container. The idea being this would only be used for the local development server, not used during package & deploy. The `start-api` command would pass this template. 

```yaml
AWSTemplateFormatVersion: '2010-09-09'
Transform: AWS::Serverless-2016-10-31
Resources:
  DevLambda:
    Type: AWS::Serverless::Function
    Properties:
      CodeUri: .
      Events:
        HttpApiProxy:
          Type: HttpApi
          Properties:
            ApiId: !Ref DevHttpApi
      Runtime: nodejs14.x
      Handler: src/index.handler
      MemorySize: 512
      Timeout: 30
  DevHttpApi:
    Type: AWS::Serverless::HttpApi
    Properties:
      StageName: development
```

```shell
sam local start-api \
  --template 'template-dev.yaml' \
  --container-host 'host.docker.internal' \
  --docker-volume-basedir $DIDPWD \
  --host '0.0.0.0' \
  --port 3031
```

Where `$DIDPWD` would be `$PWD` passed down from the host. This used to be needed during a really old SAM version but I can NOT confirm it works (it should) because...

**💥💣💥 SAM HAS ANOTHER BUG THAT PREVENTS THIS. SEE HERE:** https://github.com/aws/aws-sam-cli/issues/2837

But even if that did work, it would not be a good workaround for a container function that might have done a system dependency via yum install on both the production and development Docker files. 

## Wrapping Up

I GET IT! I get why this is hard. The `sam build` historic ways of working make it difficult to solve everything. But I do think we have been pained into a corner and exposing an interface like a watcher here is adding to the list. It would be prone to being slow. Maybe there is a better way forward?

Is there a way to use SAM's API Gateway => Lambda integration in some way to make local development work?

**Comment by metaskills** [comment]
Using @sbhvt's comment here https://github.com/aws/aws-sam-cli/issues/921#issuecomment-907859353 I was able to change my last hack above the 💥💣💥 in the following ways. First in my `docker-compose.yml` file. Then in my script used to start the server in the SAM dev container.

```dockerfile
volumes:
  - ${PWD}:${PWD}
working_dir: $PWD
```

```shell
sam local start-api \
  --template 'template-dev.yaml' \
  --container-host 'host.docker.internal' \
  --host '0.0.0.0' \
  --port 3031
```

But this only works on Mac! I can not get it to work with GitHub Codespaces yet. I tried following this advice (https://dev.to/natterstefan/docker-tip-how-to-get-host-s-ip-address-inside-a-docker-container-5anh) and used `172.17.0.1` but that did not help. 

So I'm blocked even on the hacks of hacks :(

**Comment by metaskills** [comment]
GOT IT TO WORK ON CODESPACES. Just needed to clear out Docker images and start from scratch. I'll be using `COMPOSE_FILE` to add another compose layer for Codespaces that adds this.

```dockerfile
extra_hosts:
  - host.docker.internal:172.17.0.1
```

Which allows the bin/server wrapper to do the following, the key here is was adding `--container-host-interface '0.0.0.0'` which is needed for Linux and does not hurt the Mac side of things. So final bin/server is something like this.

```shell
sam local start-api \
  --template 'template-dev.yaml' \
  --container-host 'host.docker.internal' \
  --container-host-interface '0.0.0.0' \
  --host '0.0.0.0' \
  --port 3031
```

AGAIN! This is hacks on top of hacks. IMO we do not need a watch command but better tooling to expose a simple API Gateway => Lambda server option that works with both Docker in Docker and Lambda Containers where it can be assumed that the Docker dev env has everything setup (built) where `sam build` is not needed.

**Comment by lgabrielvleon** [comment]
> Hey guys,
> 
> So, we created a small npm package based in nodemon (samwatch):
> 
> https://www.npmjs.com/package/samwatch
> 
> https://github.com/mxitgo/samwatch
> 
> The way it works is, it copies your js/json files as you save them from the source folder to the corresponding .aws-sam/build folder (as long as the names are the same for your lambda and the code uri in template.yaml).
> 
> That is useful to see your changes reflected on the fly if you are running sam local start-api, so you don't have to run sam build after every change.
> 
> On the other hand, if the corresponding file copy is not within the .aws-sam/build folder already, the package will trigger a sam build (unless you use n parameter).
> 
> The package has helped us get a hot reload feeling for our local lambda function development with sam.
> 
> I hope anyone can find it useful.
> 
> Thanks!

Hi! 

Works with PackageType: Image?

Thks!

**Comment by yuyokk** [comment]
> Workaround
> All of the sam local command allow 'reloading' of the function code. This means for a given template, if you update the contents of a functions Code/CodeUri, SAM CLI will mount this updated content. So the current option for making interactions easier is to have two terminal windows open, one that builds and one that you local invoke with.

Is there a workaround for projects created with `aws-cdk`. Do I need to run `cdk watch` or `npm run build` (ts => js) in the second terminal?

**Comment by zwkf** [comment]
Just having a sensible way to have a mounted volume inside the docker container would be awesome for my use case

**Comment by vavdoshka** [comment]
AWS, What does it take to solve that significant UX problem? 

Thanks to everyone who contributed with their solution; I'll re-iterate on what worked for me for Python project

```
# Running Locally

1. Start lambda build loop (will watch for changes and rebuild)

nodemon --exec "sam build --template template.yaml --use-container" -e py
* install nodemon globally if you don't have it already `npm install -g nodemon`

2. Start local api server on port 5000

sam local start-api -p 5000
```

**Comment by S33G** [comment]
> AWS, What does it take to solve that significant UX problem?
> 
> Thanks to everyone who contributed with their solution; I'll re-iterate on what worked for me for Python project
> 
> ```
> # Running Locally
> 
> 1. Start lambda build loop (will watch for changes and rebuild)
> 
> nodemon --exec "sam build --template template.yaml --use-container" -e py
> * install nodemon globally if you don't have it already `npm install -g nodemon`
> 
> 2. Start local api server on port 5000
> 
> sam local start-api -p 5000
> ```

This appears to be a JavaScript example

But Python would need to be handled also

**Comment by vavdoshka** [comment]
> > AWS, What does it take to solve that significant UX problem?
> > Thanks to everyone who contributed with their solution; I'll re-iterate on what worked for me for Python project
> > ```
> > # Running Locally
> > 
> > 1. Start lambda build loop (will watch for changes and rebuild)
> > 
> > nodemon --exec "sam build --template template.yaml --use-container" -e py
> > * install nodemon globally if you don't have it already `npm install -g nodemon`
> > 
> > 2. Start local api server on port 5000
> > 
> > sam local start-api -p 5000
> > ```
> 
> This appears to be a JavaScript example
> 
> But Python would need to be handled also

Nope, this works in python, it watches the files with the .py extension - `-e py`

After playing with it for some time, I've added a delay to avoid re-building too quickly.

`nodemon --exec "sam build --template template.yaml --use-container" -e py --delay 2.5`

**Comment by cesar-presence** [comment]
almost 4 years of this issue. Any news?

**Comment by DanielKucal** [comment]
Here's the command I ended up with for now:
```
sam build && concurrently "nodemon --on-change-only --ext ts --exec sam build" "sam local start-api"
```
Requires `npm i -g nodemon concurrently`

It runs nodemon and `sam local start-api` in the background simultaneously. The trick is to have the build available before creating docker image, so it's done beforehand and nodemon does it only after changes (using `--on-change-only` flag).

**Comment by paul-uz** [comment]
+1 for this feature

**Comment by wyatt-shein** [comment]
I've only been experiencing this same issue for a bit so im only vaguely familiar but id like to a propose another solution that the sam team at AWS may be able to implement more easily than creating hot reload for the .aws-sam built folder:

Context:
Im using node but you can do this with others like python in a similar way.
If i have a lambda layer that manages my dependencies and some lambda functions that utilize the dependencies in that layer, I can set my folders up in the following way that allows me to use my functions successfully with hot reload and without having to run sam build.

Article: https://aws.amazon.com/blogs/compute/working-with-aws-lambda-and-lambda-layers-in-aws-sam/
$ tree sam-app
sam-app
├── README.md
├── dependencies
│   └── nodejs
│       └── package.json
├── hello-world
│   ├── app.js
│   └── tests
│       └── unit
└── template.yaml

When using the nodejs folder as described in the article and the docs, as long as you have generated the node modules manually using npm i, it should work just fine to run your code locally and have hot reloading without using sam build. 

The issue with this method is that when you need to move beyond local environments to testing or production, most pipelines will utilize sam build. When sam build is run, it will generate a nodejs folder inside your nodejs folder and not generate the dependencies correctly as it would if you didnt have the nodejs folder created (discussed in previous comments). 

So what im saying is maybe the sam developers would have an easier time fixing the current issue of duplicate nodejs folders which causes the dependency generation to not work, rather than trying to fix the issue of hot reload not working for the 'sam build' built .aws-sam folder.

**Comment by Frostie2** [comment]
Coming here quite late but here's an awesome workaround (unix systems only, sorry Windows users, I'm sure there's an equivalent):

`find . -not -path "./.*" | entr -r sh -c 'sam build && sam local start-api'`

Whenever we change anything in the source code it automatically rebuilds and starts the api. Works like a charm and saves me so much time having to manually rebuild. Now I don't even need to look.

**Explanation:**

- `find .` finds all files in the current directory
- `-not -path "./.*"` ensures we're not tracking the build and any other folders starting with "." as this can cause recursive builds.
- We pipe all these files into `entr` which watches for changes on each of these files.
- `entr -r` makes sure the process created by the command we specify after is killed / restarted on file changes.
- `sh -c <command>` runs the command, in this case: build and start the api

**Requirements:**
Install `entr` using
`brew install entr`

You're good to go!

**Comment by yjo** [comment]
As another workaround when using TypeScript/esbuild that doesn't require re-running the SAM CLI I've added to `package.json` a scripts entry like:
```
"watch": "esbuild --watch app.ts --bundle --platform=node --outdir=../.aws-sam/build/MyFunction --format=cjs --target=es2020 --sourcemap"
```
And I can now `npm run watch` while `sam local start-api` is running to have SAM recreate the function container as needed. You can check the output of `sam build --debug` to see the arguments it passes to `esbuild` and tweak accordingly.

**Comment by dcabib** [comment]
## Hey there! 👋 Found a solid path forward for implementing `--watch`

So I dug deep into the SAM CLI codebase and honestly, we're like 80% of the way there already. The infrastructure for file watching is already built and battle-tested - it's what powers `sam sync --watch`.

### What's already there

The codebase has all the heavy lifting done:

- `WatchManager` class that handles the file watching logic
- `HandlerObserver` using `watchdog` to actually monitor files
- `ResourceTrigger` classes that know how to detect changes in different resource types
- All the thread safety and graceful shutdown stuff

Basically, `sam sync --watch` already does 90% of what we need, just for a different purpose.

### The implementation would be pretty straightforward

Just need to:

1. __Add the `--watch` flag__ to `sam build` (literally like 5 lines of code)

2. __Create a `BuildWatchManager`__ that copies the same patterns from `SyncWatchManager` but calls `build_context.run()` instead of sync operations

3. __Wire it up__ in the build command - if `--watch` is passed, start the watch manager, otherwise do the normal single build

### What it would look like

```bash
sam build --watch                    # watches everything
sam build MyFunction --watch         # watches just that function
sam build --watch --use-container    # works with all existing flags
```

When files change, it just rebuilds and shows you what happened. Clean, simple.

### Why this approach rocks

- We're not reinventing anything - just reusing proven code
- It'll work with containers, caching, parallel builds, whatever
- No weird workarounds needed anymore
- Same UX patterns as the rest of SAM CLI

### The effort

Honestly not that much work. Maybe a couple hundred lines of code max, mostly copy-pasting patterns that already work. The hard stuff (file watching, change detection, threading) is already done.

I've been working with this codebase and I'm confident this would be clean and maintainable. Happy to take a stab at implementing it if the maintainers are cool with this approach.

Thoughts?

**Comment by roger-zhangg** [comment]
@dcabib Thanks for the solution! I appreciate the analysis of leveraging the existing `sam sync --watch` infrastructure.

My main concern is the potential for recursion issues. For instance, with flags like `--build-in-source` or when using `--cache-dir`, the watcher might end up monitoring directories that contain build outputs, which could trigger infinite rebuild loops.

I'm assuming that `sam build --watch` would work cleanly only when all build artifacts are contained within the `.aws-sam` directory (the default behavior). This way, the watcher can simply exclude `.aws-sam` and avoid watching its own outputs. But when build artifacts end up elsewhere, things could get tricky.

Have you considered how to handle these edge cases?

**Comment by dcabib** [comment]
@roger-zhangg I did analyze how `sam sync --watch` handles this, and there's already a solid reference we can reuse part of this logic. The existing `CodeResourceTrigger` classes have a robust exclusion system with `DEFAULT_WATCH_IGNORED_RESOURCES` that already excludes `.aws-sam` directories and `node_modules`.

For the specific edge cases you mentioned:

__`--build-in-source`__: We can extend the exclusion patterns dynamically based on the build configuration. When this flag is used, we'd exclude common build artifact patterns (`.pyc`, `__pycache__`, `target/`, `build/`, etc.) from the source directories being watched.

__Custom `--cache-dir`__: What about detect when the cache directory falls under the base project directory and automatically add it to the exclusion list? If it's outside the project directory, no exclusion needed

The key insight is that we don't need to prevent these flag combinations entirely - we just need to be smart about what we exclude from watching based on the build configuration. The `WatchManager` already accepts a `watch_exclude` parameter that we can populate intelligently.

Since `sam sync --watch` already handles similar complexity with exclusions and hasn't had recursion issues, I'm kind of confident we can make `sam build --watch` equally robust using the same patterns.

Would this approach address your concerns?

**Comment by dcabib** [comment]
## 🎉 Issue #921 - SOLUTION IMPLEMENTED!

Hi @roger-zhangg and AWS SAM team - I am excited to report that the 6-year-old `sam build --watch` request has been **fully implemented and tested**! 

## 🔍 **Discovery:**
While analyzing the codebase, I discovered that `sam build --watch` was **already implemented** but had two critical bugs:
1. **Help Discoverability**: The `--watch` option was not visible in help text
2. **Template Change Detection**: BuildContext was using stale template data

## 🛠️ **Fixes Applied:**
- ✅ **Help Visibility**: Added "watch" to `BUILD_STRATEGY_OPTIONS` - option now appears in help
- ✅ **Template Detection**: Fixed by calling `build_context.set_up()` before `build_context.run()` 
- ✅ **Enhanced Template Monitoring**: Improved file watching with proper callbacks

## 📊 **Comprehensive Testing Results:**
- ✅ **Source Code Changes**: Python, Node.js files → Instant auto-rebuild
- ✅ **Template Changes**: Adding new functions → Auto-rebuild  
- ✅ **Dependencies**: requirements.txt, package.json → Auto-rebuild
- ✅ **Container Builds**: Works with `--use-container` for complex dependencies
- ✅ **Multi-Runtime**: Python, Node.js, Java, Ruby, .NET, Go
- ✅ **Performance**: Debouncing, caching, parallel builds
- ✅ **Error Recovery**: Continues watching after build failures

## 🚀 **Ready for Review:**
**PR #8286**: https://github.com/aws/aws-sam-cli/pull/8286

This **completely eliminates** the need for:
- ❌ Manual `sam build` after every code change
- ❌ Community workarounds (nodemon, samwatch, webpack scripts)
- ❌ Two terminal setup (build in one, run in another)

## 💻 **Usage:**
```bash
sam build --watch                    # Basic projects
sam build --watch --use-container    # With dependencies  
sam build MyFunction --watch         # Single function
sam build --watch --cached           # With caching
sam build --watch --parallel         # Parallel builds
```

After 6 years and 25+ frustrated community comments, developers can finally enjoy **seamless local development** with automatic rebuilds! 🎉

**This addresses every single pain point mentioned in the thread** - from @alxx Ruby gems to @AmillerSC minor changes to @anubhavmalik console.log() edits.

Ready for review and community testing!

**Comment by alxx** [comment]
Wow, and it's only been 6 years :) now if I could only remember what AWS SAM was...
